File: Data Structure/4. Stack/arrayStack.py
# simple implementation with list
class ArrayStack:
  # constructor
  def __init__(self):
    self.stack = [];

  # check if the stack is empty
  def is_empty(self):
    return len(self.stack) == 0

  # peek the top of the stack
  def peek(self):
    # if empty
    if self.is_empty():
      print("The stack is empty!")
    return self.stack[-1] # not empty, peek it

  # push value into the stack
  def push(self, value):
    self.stack.append(value)
  
  # pop the top of the stack
  def pop(self):
    # if empty
    if self.is_empty():
      print("The stack is empty!")
      return
    self.stack.pop() # pop if not empty

  # size of the stack
  def size(self):
    return len(self.stack)
  
  # output the stack
  def output(self):
    if self.is_empty():
      print("stack is empty!")
    print("contents: "),
    print(self.stack[::-1])
    print

# ============================
# testing the stack class
# ============================

print("\nList implementation of stack!\n")

# create a new ArrayStack object
stack = ArrayStack()
print("- New ArrayStack object created")
stack.output()

# push value into the stack
stack.push(1)
stack.push(2)
stack.push(3)
print("- New values are pushed into the stack")
stack.output()

# peek value
print("- Peek the top of the stack: " + str(stack.peek()))
print

# pop the top of the stack
stack.pop()
print("- Pop the top of the stack")
stack.output()

# get size of the stack
print("- Size of the stack: " + str(stack.size()))
print

File: Data Structure/4. Stack/arrayStack.cpp
// a simple implementation with vector
#include <iostream>
#include <vector>
using namespace std;

class ArrayStack {
  public:
  vector<int> stack;
  int peek() {return stack.back();} // peek the top of the stack
  int isEmpty() {return stack.empty();} // check if the stack is empty
  int size() {return stack.size();} // return the size of the stack
  void push(int value) {stack.push_back(value);} // push value to the top
  void pop() {if (!isEmpty()) stack.pop_back();} // pop recent value
  void output() {
    // empty, no need for output
    if (stack.empty()) {
      cout << "stack is empty!" << endl;
      return;
    }

    // print out the values
    cout << "contents: ";
    vector<int> copy = stack;
    while (!copy.empty()) {
      cout << copy.back() << " ";
      copy.pop_back();
    }
    cout << endl;
  }
};

// print function
void print(string text) {cout << text;}

// test the stack
int main() {
  print("Array implementation of stack")

  // create a stack
  ArrayStack stack;
  print("-stack created!\n");
  stack.output(); // output if content exists
  print("\n");

  // push value into the stack
  stack.push(1);
  stack.push(2);
  stack.push(3);
  print("-values are pushed into the stack!\n");
  stack.output(); // output if content exists
  print("\n");

  // pop value from the top
  stack.pop();
  print("-pop the stack!\n");
  stack.output(); // output if content exists
  print("\n");

  // size of the staclk
  cout << "-The size is " << stack.size() << "\n";
  stack.output(); // output if content exists
  print("\n");
}